Custom components, plus some tweaks

psndcsrv opened the pull request:
New Components:
* Color Indicator
  * Takes a value and converts it to a color and displays it in a square
* Joystick
  * Drag a "handle" within a base to output a value with a magnitude and direction

Modified Components:
* Sliders
  * now support having their titles below, or inline to the left or right
  * now support a fill color drawn from the handle to a specific value on the slider
* Div components
  * now support multiline content specified in an array, which gets concatenated (similar to the way scripts work)

psndcsrv commented:
@pjanik Would you mind reviewing these changes when you get a chance? This will likely conflict with @rmarren1 's work in pull #80, so don't worry about merging this -- I can resolve any conflicts after that one is merged. Just let me know if you think anything needs reworked.

Thanks!

pjanik commented on src/lab/common/controllers/color-indicator-controller.js:
```diff
@@ -0,0 +1,141 @@
+          // emSize = parseFloat($sliderHandle.css("font-size"));
+          // $sliderHandle.css("height", remainingHeight + emSize * 0.4);
+          // $sliderHandle.css("top", -0.5 * remainingHeight - emSize * 0.4);
+        // }
```
I guess these comments aren't relevant to this component. ;)

psndcsrv replied:
Ah, nope. Cruft from copying from the slider component that I commented rather than deleting. I'll strip them out.

pjanik commented on src/lab/common/controllers/interactive-metadata.js:
```diff
@@ -818,6 +823,103 @@ define(function() {
+      title: {
+        defaultValue: ""
+      },
+      colorValue: {
```
At the beginning I didn't know what `colorValue` was and what I should have put there, so some comment with example would be great.

pjanik commented on src/lab/common/controllers/color-indicator-controller.js:
```diff
@@ -0,0 +1,141 @@
+          // Update "initialValue" to represent current
+          // value of the slider.
+          result.initialValue = 0;
+        }
```
Probably code above shouldn't be there.

pjanik commented:
Frankly speaking, I'm not a big fan of the "class" pattern that is used by slider-component (and these new components which were probably based on it). I was hoping that we could slowly move towards to classic, prototype-based approach, e.g. like this one:
https://github.com/concord-consortium/lab/blob/master/src/lab/common/controllers/button-controller.js
But I guess it doesn't makes sense to change it now unless there is some benefit from using `InteractiveComponent` as a base class.

I tried to add these components to this interactive:
http://lab.concord.org/interactives.html#interactives/basic-examples/interactive-components.json
Color indicator worked, but joystick didn't. I provided totally minimal set of options (id and type, then tried initialValue), so probably that's the culprit. But it would be nice to mark some options as required or provide default values, so it always works when the component definition is correct (even if it's minimal). And maybe some hight level description of joystick and its "value", because it's difficult to guess what will be assigned to the model property if I bind joystick to it.

pjanik commented on src/lab/common/controllers/interactive-metadata.js:
```diff
@@ -818,6 +823,103 @@ define(function() {
+        // {
+        //   "value": [value, e.g. 100],
+        //   "label": [label, e.g. "High"]
+        // }
```
Looks like slider comments. :wink:

pjanik commented on src/lab/common/controllers/slider-controller.js:
```diff
@@ -71,6 +73,66 @@ define(function () {
+          $container.css('background', '-moz-linear-gradient(left, ' + gradientStr + ')');
+          $container.css('background', '-o-linear-gradient(left, ' + gradientStr + ')');
+          $container.css('background', 'linear-gradient(to right, ' + gradientStr + ')');
+          console.log("Setting gradient: linear-gradient(to right, " + gradientStr + ")");
```
Debugging leftover?

psndcsrv commented:
@pjanik I just pushed a commit which addresses your comments. I think the totally minimal definition for the joystick should work now, too.

I did use the slider as a basis for my new components. I don't really have strong opinions either way about classic vs class based approaches, so if we want to refactor and change them, that'd be ok. I'm on other projects now, though, so it'd have to wait for a future iteration if we did.

pjanik commented:
Thanks! Unfortunately I still can't add joystick to the interactive, as Lab is failing:

![screen shot 2015-04-20 at 12 40 57](https://cloud.githubusercontent.com/assets/767857/7228566/a78d00ca-e75a-11e4-95f9-bea21e4e7a18.png)

There is one more thing that came to my mind - I'm not 100% sure now, but I think that Lab properties should be restricted to primitive values (numbers, strings).

I remember we used to have problems because someone was using an array as property (or parameter) value. However I can't remember now what was the issue - property caching, update notification or tick history? I also remember that we tried to workaround that, but I'm not sure now what's the current status. @rklancer, don't you remember by chance?

I would test if `onPropertyChange` and tick history work as expected with joystick. If so, I think it's fine to leave it for now. Otherwise, I guess we would have to divide joystick's `property` into two separate properties (e.g. `magnitudeProperty` and `distanceProperty`).

scytacki commented:
Unless it is a lot of work, I think it is best to switch to use two separate properties. I think this will also help with logging when we get it working.
@psndcsrv do you think you can fit that in?

scytacki commented:
@rklancer is going to be working on HAS Lab changing during the next week. So I think the work to separate out the properties can be done by him, unless it is difficult for some reason.

psndcsrv commented:
@scytacki No, it should be straightforward. I had originally just put it in the same property, since the parts are kind of useless without each other. But if it's going to cause problems, then yeah, we can split it easily enough.

rklancer commented:
@pjanik is correct, Lab properties really should be primitive values. You can get away with reference types (objects & arrays) but that's usually a bad idea because change detection won't happen when you mutate the object/array without replacing it. So script authors would always have to remember to do something like

    wind = { magnitude: wind.magnitude * 1.1, direction: wind.direction };

and never to do

    wind.magnitude *= 1.1;

because that would not update the joystick, and properties that depend on wind would not be updated.

rklancer commented:
And yet ... it's not as simple as that! There's a strong assumption in Lab that components are bound to a _single_ value. Let me think a bit...

rklancer commented:
Okay. @scytacki and @pjanik, tell me what you think. We should _keep_ the compound value, but we can fix

pjanik commented:
> There's a strong assumption in Lab that components are bound to a single 'value' property.

Out of curiosity, why?

rklancer commented:
Okay. @scytacki, @pjanik, @psndcsrv, tell me what you think of this:

* We should _keep_ the compound property `value`, because component serialization logic and similar concerns (such as `retainProperty`) expect that a component is bound to a single property (if any)
* To handle the resulting problems, we create an Github issue to update property support so that things work correctly when the `value` property is mutated (e.g., when the `value`'s `magnitude` property is changed without changing the object reference)

I believe that the latter can be handled by a three small changes to propertySupport.

* Where we [return the value to save into tick history](https://github.com/concord-consortium/lab/blob/554be905bfac7d2067708a9e76cdf073f8a582df/src/lab/common/property-support.js#L933) we could make a clone.
* When we [set a property](https://github.com/concord-consortium/lab/blob/554be905bfac7d2067708a9e76cdf073f8a582df/src/lab/common/property-support.js#L493), check to see if it's a primitive value, and if it's not, endow the value with a simple `toString` like method that returns a meaningful string or hash that can be used to check for equality (e.g. it could return a string like `obj: a:1,b:2` for `{a: 1, b: 2}`)
* Where we [check for changed property values](https://github.com/concord-consortium/lab/blob/554be905bfac7d2067708a9e76cdf073f8a582df/src/lab/common/property-support.js#L870) we use the above-provided method to check for equality if the type is not a primitive type.

I probably don't have the time right now to be totally sure that this works and doesn't cause any regressions, and I think the desired interactive works OK without these changes, which is why I'm proposing to leave it as is for now and track the fix with an issue.

pjanik commented:
I think that it would be generally useful to support compound values. It's not clear that they are not supported and it can bite us again in the future. The question is if we have time to implement it. :wink:

So, if we always use in `property = {key1: newValue, key2: oldValue}` instead of `property.key1 = newValue` in scripts, will everything work fine for now? I mean property support, tick history, etc.?

> When we set a property

Hm, but what when we do `property.key = value;`? It might be difficult to detect a change (unless we wrap every value in custom object with special setters).

Regarding `toString` idea, I would also consider cloning and checking if objects are equal to each other. Then just pick one which is faster in practice for typical data (small objects, small arrays).

rklancer commented:
@pjanik I suppose it's not a "strong" assumption. It's been a _convention_ that interactive components may have a `property` value, but at some point, the `retainProperty` property was added, and that implementation does assume that components have a property called `property`: https://github.com/concord-consortium/lab/blob/9b11d876937e26736bb61857f4aab17a823b2762/src/lab/common/controllers/interactives-controller.js#L383

It seems worthwhile to maintain that convention, given that authors and developers are likely to assume it and devs. Also, because I do think the changes to make compound properties (at least simple hashes and arrays) work correctly are minor, I think it's better to make that improvement than to make the joystick controller violate the existing convention.

Note that it might not have been obvious that we could make that improvement! If computed property support depended on _observing_ property setting, it would be hard to make compound properties work correctly. But it depends on batch update because of the physics engine, and that makes it easier to support computed properties that depend on the value of compound properties.

rklancer commented:
One comment re the above scheme for compound property support is that nothing will happen if you mutate a compound property when the model is not running. But already script authors have to know to wrap things in `makeInvalidatingChange`, because it's already the case that you need to do that for computed properties to work if they depend on properties that are changed outside of a model tick.

rklancer commented:
@pjanik yes, I'm pretty sure that it should work. As I noted, if you do `property.key = x` during a model tick (ie in a script) things should just work. (The change will be detected, causing controls to be notified, and the values put into the tick history will always be clones). When you're somehow not in a tick -- for example, if Lab code is manipulating a property in this way -- you will specifically have to mark it invalidating.

For what it's worth, I think the time commitment lies more in testing and double checking that compound values work as expected, rather than in the changes required to make it work.

I believe the joystick works OK when used specifically with the AgentScript iframe model, where for now it is intended to be used. 

But, yes, I'm proposing merging this pull request and ignoring the fact that it relies on compound-value properties that aren't perfectly supported yet.

rklancer commented:
Pivotal Tracker story: https://www.pivotaltracker.com/n/projects/442903/stories/93559672
